internal/armada/cache/queue.go

```diff
@@ -0,0 +1,155 @@
+package cache
+
+import (
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/G-Research/armada/internal/armada/repository"
+	"github.com/G-Research/armada/internal/armada/scheduling"
+	"github.com/G-Research/armada/internal/common"
+	"github.com/G-Research/armada/pkg/api"
+)
+
+type empty struct{}
+type stringSet map[string]empty
+
+type QueueCache struct {
+	queueRepository          repository.QueueRepository
+	jobRepository            repository.JobRepository
+	schedulingInfoRepository repository.SchedulingInfoRepository
+
+	refreshMutex           sync.Mutex
+	queuedResources        map[string]map[string]common.ComputeResourcesFloat
+	queueNonMatchingJobIds map[string]map[string]stringSet
+}
+
+func NewQueueCache(
+	queueRepository repository.QueueRepository,
+	jobRepository repository.JobRepository,
+	schedulingInfoRepository repository.SchedulingInfoRepository,
+) *QueueCache {
+	collector := &QueueCache{
+		queueRepository:          queueRepository,
+		jobRepository:            jobRepository,
+		schedulingInfoRepository: schedulingInfoRepository,
+		queuedResources:          map[string]map[string]common.ComputeResourcesFloat{},
+		queueNonMatchingJobIds:   map[string]map[string]stringSet{}}
+
+	return collector
+}
+
+func (c *QueueCache) Refresh() {
+	queues, e := c.queueRepository.GetAllQueues()
+	if e != nil {
+		log.Errorf("Error while getting queue metrics %s", e)
+		return
+	}
+
+	clusterInfo, e := c.schedulingInfoRepository.GetClusterSchedulingInfo()
+	if e != nil {
+		log.Errorf("Error while getting cluster reports %s", e)
+		return
+	}
+
+	activeClusterInfo := scheduling.FilterActiveClusterSchedulingInfoReports(clusterInfo)
+	clusterInfoByPool := scheduling.GroupSchedulingInfoByPool(activeClusterInfo)
+
+	for _, queue := range queues {
+		resourceUsageByPool := map[string]common.ComputeResources{}
+		nonMatchingJobs := map[string]stringSet{}
+
+		err := c.jobRepository.IterateQueueJobs(queue.Name, func(job *api.Job) {
+			jobResources := common.TotalJobResourceRequest(job)
+			nonMatchingClusters := stringSet{}
+
+			for pool, infos := range clusterInfoByPool {
+				matches := false
+				for _, schedulingInfo := range infos {
+					if scheduling.MatchSchedulingRequirements(job, schedulingInfo) {
+						matches = true
+					} else {
+						nonMatchingClusters[schedulingInfo.ClusterId] = empty{}
+					}
+				}
+
+				if matches {
+					r, exists := resourceUsageByPool[pool]
+					if !exists {
+						r = common.ComputeResources{}
+						resourceUsageByPool[pool] = r
+					}
+					r.Add(jobResources)
+				}
+			}
+			nonMatchingJobs[job.Id] = nonMatchingClusters
+		})
+
+		if err != nil {
+			log.Errorf("Error while getting queue %s resources %s", queue.Name, err)
+		}
+
+		c.updateQueuedNonMatchingJobs(queue.Name, nonMatchingJobs)
```

Comment:
Maybe better to update these 2 in the same mutex?

Reply:
This way the locking is cleaner in update methods, the updates are not frequent enough to justify making it more complicated.